applications/StructuralMechanicsApplication/custom_response_functions/response_utilities/eigenfrequency_response_function_utility.h

```diff
@@ -290,33 +287,37 @@ class EigenfrequencyResponseFunctionUtility
             for(auto& node_i : elem_i.GetGeometry())
             {
                 Vector gradient_contribution(3, 0.0);
-                Matrix perturbed_LHS = Matrix(num_dofs_element,num_dofs_element);
-                Matrix perturbed_mass_matrix = Matrix(num_dofs_element,num_dofs_element);
+                Matrix derived_LHS = Matrix(num_dofs_element,num_dofs_element);
+                Matrix derived_mass_matrix = Matrix(num_dofs_element,num_dofs_element);
 
                 for(std::size_t coord_dir_i = 0; coord_dir_i < domain_size; coord_dir_i++)
                 {
-                    node_i.GetInitialPosition()[coord_dir_i] += mDelta;
-                    node_i.Coordinates()[coord_dir_i] += mDelta;
-
-                    elem_i.CalculateMassMatrix(perturbed_mass_matrix, CurrentProcessInfo);
-                    noalias(perturbed_mass_matrix) = (perturbed_mass_matrix - mass_matrix_org) / mDelta;
-
-                    elem_i.CalculateLocalSystem(perturbed_LHS, dummy ,CurrentProcessInfo);
-                    noalias(perturbed_LHS) = (perturbed_LHS - LHS_org) / mDelta;
+                    if( coord_dir_i == 0 )
+                    {
+                        DifferentiationUtility::CalculateLeftHandSideDerivative(elem_i, SHAPE_X, node_i, mDelta, derived_LHS, CurrentProcessInfo);
```

Comment:
here you could create a vector of `[SHAPE_X, SHAPE_Y, SHAPE_Z]` and iterate over that instead of the if/else

Reply:
Or maybe it is even possible to get the component from the main variable `SHAPE` itself, using an index?

Reply:
I will try it